### PR TITLE
Remove (.[]<-)

### LINKED ISSUE #11342: "str.[i] <- c" is desugared into "String.set str i c" which does not exist anymore
Using the `5.0` branch (https://github.com/ocaml/ocaml/commit/f5d778a335c653da33c1c021a445fd910198089c), the following can be seen in the REPL:
```
# "x".[0] <- 'y';;
Error: Unbound value String.set
Hint: Did you mean get?
# (Bytes.create 1).[0] <- 'y';;
Error: Unbound value String.set
Hint: Did you mean get?
```

The first expression `"x".[0] <- 'y'` is desugared into `String.set "x" 0 'y'`, however `String.set` does not exist anymore since #10867. To me, this expression should raise a type error on `"x"` being `string` instead of the expected `bytes`.

The second expression `(Bytes.create 1).[0] <- 'y'` is desugared into `String.set (Bytes.create 1) 0 'y'`, however, once again, `String.set` does not exist anymore. To me this should instead desugar to `Bytes.set (Bytes.create 1) 0 'y'` which would be correct.

### PR DESCRIPTION
**Pull request by nojb**
This operator should have been removed when the function `String.set` was removed in #10867.

Fixes #11342

### CODE REVIEW

**Comment by xavierleroy** [comment]
The only potential downside I see to this change is that it rules out the following valid OCaml 4 code:
```
module BytesNotation = struct
    module String =
      let get = Bytes.get
      let set = Bytes.set
    end
  end;;
# let b = Bytes.make 25 'x';;
val b : bytes = Bytes.of_string "xxxxxxxxxxxxxxxxxxxxxxxxx"
# BytesNotation.(b.[2] <- 'y');;
- : unit = ()
# BytesNotation.(b.[2]);;
- : char = 'y'
```
I don't really want support this kind of abuse of notations, however.

**Comment by Octachron** [comment]
Do we want a specialized error message? It is quite exceptional that the removal of a deprecated function might result in a syntax error in user code. It might be help confused users to have a specific explanation rather than just a `Error: Syntax error` located on the assignment arrow `<-` .

**Comment by xavierleroy** [comment]
I agree that a specialized error message would be nice, at least in 5.0.  It could be removed in later releases.

**Comment by nojb** [comment]
> I agree that a specialized error message would be nice, at least in 5.0. It could be removed in later releases.

I gave this a try, it looks like this:

```
$ TERM=dumb ocaml
OCaml version 5.1.0+dev1-2022-06-09
Enter #help;; for help.

# s.[i] <- x;;
Line 1, characters 0-10:
1 | s.[i] <- x;;
    ^^^^^^^^^^
Error: the infix operator (.[]<-) used to be bound to String.set, which has been removed in OCaml 5.0
```

**Comment by dbuenzli** [comment]
> Error: the infix operator (.[]<-) used to be bound to String.set, which has been removed in OCaml 5.0

The message doesn't say what the problem is. You should say it is illegal syntax.

**Comment by nojb** [comment]
> > Error: the infix operator (.[]<-) used to be bound to String.set, which has been removed in OCaml 5.0
> 
> The message doesn't say what the problem is. You should say it is illegal syntax.

OK, I tweaked the message:

```
$ TERM=dumb ocaml
OCaml version 5.1.0+dev1-2022-06-09
Enter #help;; for help.

# s.[i] <- c;;
Line 1, characters 0-10:
1 | s.[i] <- c;;
    ^^^^^^^^^^
Error: Syntax error: the infix operator (.[]<-) used to be bound to String.set, which was removed in OCaml 5.0
```

File: testsuite/tests/parsetree/source.ml [hunk 1]
```diff
@@ -5052,10 +5052,10 @@ let get_buf s i =
 let set_buf s i u =
   let n = UChar.uint_code u in
   begin
-    s.[i] <- Char.chr (n lsr 24);
-    s.[i + 1] <- Char.chr (n lsr 16 lor 0xff);
-    s.[i + 2] <- Char.chr (n lsr 8 lor 0xff);
-    s.[i + 3] <- Char.chr (n lor 0xff);
+    s.![i] <- Char.chr (n lsr 24);
```
Comment (kit-ty-kate) [inline testsuite/tests/parsetree/source.ml:5055]: I’m a bit surprised by this change. Where is the `(.![]<-)` operator defined? I wasn’t able to locate it

Comment (nojb) [reply testsuite/tests/parsetree/source.ml:5055]: This test only checks that the source file can be parsed.

Comment (kit-ty-kate) [reply testsuite/tests/parsetree/source.ml:5055]: ah I see. Thanks!

**Comment by Octachron** [comment]
I am not that fond of having history-focused user messages. When reading 
 ```
  Syntax error: the infix operator (.[]<-) used to be bound to String.set, which was removed in OCaml 5.0
```
a new user will have to search what was `String.set` in older versions of OCaml, understand why it existed, then why it was removed before finally making the link with their current issue.

Even regular users might scratch their head about the information that `.[]<-` was bound to `String.set` (or `String.unsafe_set`) which was always an implementation detail.

I would propose to focus on the current situation:
- strings are immutable
- an assignment operator makes no sense for immutable data types
- there is a mutable sequence of bytes data types: Bytes.t

Something like
```
Syntax error, strings are immutable, there are no assignment syntax for them.
Hint: Mutable sequence of bytes are available in the Bytes module.
```
but I am missing a good way to bring up `Bytes.set`. And there is some potential confusion for people that knowingly ignored the deprecation warning for `.[]<-` and got used to the idea that `.[]<-` worked on bytes rather than strings.

**Comment by nojb** [comment]
> ```
> Syntax error, strings are immutable, there are no assignment syntax for them.
> Hint: Mutable sequence of bytes are available in the Bytes module.
> ```

Thanks, this feels much better. I amended the patch:

```
OCaml version 5.1.0+dev1-2022-06-09
Enter #help;; for help.

# s.[i] <- c;;
Line 1, characters 0-10:
1 | s.[i] <- c;;
    ^^^^^^^^^^
Error: Syntax error: strings are immutable, there is no assignment syntax for them.
Hint: Mutable sequences of bytes are available in the Bytes module.
#
```

**Review by xavierleroy — APPROVED** [review]
Looks good to me.

**Comment by nojb** [comment]
> Looks good to me.

Thanks. I rebased on `trunk` (with the Changes entry on 5.0 section). After merging I will cherry-pick to the 5.0 branch.

**Comment by gasche** [comment]
```
Hint: Mutable sequences of bytes are available in the Bytes module.
Instead of `s.[i] <- c`, one may write `Bytes.set buf i c`.
```

(I'm not suggesting to pretty-print the actual subterms of the user, that sounds like a dubious idea, but I think showing some concrete syntax can help.)

**Comment by nojb** [comment]
> ```
> Hint: Mutable sequences of bytes are available in the Bytes module.
> Instead of `s.[i] <- c`, one may write `Bytes.set buf i c`.
> ```
> 
> (I'm not suggesting to pretty-print the actual subterms of the user, that sounds like a dubious idea, but I think showing some concrete syntax can help.)

Personally feels a bit overkill to me, but am not opposed. If no-one voices an objection, I'll amend accordingly.

**Comment by gasche** [comment]
I think that for beginners there is a certain distance between "use the module Bytes" and (1) finding the Bytes documentation then (2) figuring out that Bytes.set is the right function and (3) how to use it. Seeing the concrete code can save a bit of time.

**Comment by Octachron** [comment]
I agree that we want to mention `Bytes.set`. Do we need to refer to the previous syntax, however
```
Hint: Mutable sequences of bytes are available in the Bytes module.
Hint: Did you mean to use `Bytes.set buffer index char`?
```
?

**Comment by nojb** [comment]
> I think that for beginners there is a certain distance between "use the module Bytes" and (1) finding the Bytes documentation then (2) figuring out that Bytes.set is the right function and (3) how to use it. Seeing the concrete code can save a bit of time.

OK, I pushed a commit as suggested.

**Comment by nojb** [comment]
> Hint: Did you mean to use `Bytes.set buffer index char`?

What about

```
Hint: Did you mean to use 'Bytes.set'?
```
?
(otherwise I'm afraid that `buffer`, `index`, `char` may end up confusing rather than clarifying)

**Comment by Octachron** [comment]
Yes, your version seems clearer to me.

**Comment by dbuenzli** [comment]
> (otherwise I'm afraid that `buffer`, `index`, `char` may end up confusing rather than clarifying)

Agreed. If you don't reply with the names that are in the context it is going to be confusing.

**Comment by nojb** [comment]
OK, pushed a new version:

```
nojebar@PERVERSESHEAF:~/ocaml$ TERM=dumb runtime/ocamlrun ./ocaml -I stdlib
OCaml version 5.1.0+dev1-2022-06-09
Enter #help;; for help.

# s.[i] <- c;;
Line 1, characters 0-10:
1 | s.[i] <- c;;
    ^^^^^^^^^^
Error: Syntax error: strings are immutable, there is no assignment syntax for them.
Hint: Mutable sequences of bytes are available in the Bytes module.
Hint: Did you mean to use 'Bytes.set'?
```

**Comment by gasche** [comment]
I'm okay with the proposed final version.

**Comment by nojb** [comment]
Cherry-picked to 5.0: 67863ba31064aed95729b93edbb09d8a7235d40a

**Pull request merged**